ray-operator/test/e2erayservice/rayservice_upgrade_test.go

```diff
@@ -0,0 +1,117 @@
+
+	rayService, err := test.Client().Ray().RayV1().RayServices(namespace.Name).Apply(test.Ctx(), rayServiceAC, TestApplyOptions)
+	g.Expect(err).NotTo(HaveOccurred())
+	g.Expect(rayService).NotTo(BeNil())
+
+	test.T().Logf("Waiting for RayService %s/%s to running", rayService.Namespace, rayService.Name)
+	g.Eventually(RayService(test, rayService.Namespace, rayService.Name), TestTimeoutMedium).
+		Should(WithTransform(RayServiceStatus, Equal(rayv1.Running)))
+
+	// Get the latest RayService
+	rayService, err = GetRayService(test, namespace.Name, rayServiceName)
+	g.Expect(err).NotTo(HaveOccurred())
+	g.Expect(rayService).NotTo(BeNil())
+
+	// Create curl pod
+	curlPodName := "curl-pod"
+	curlContainerName := "curl-container"
+
+	curlPod, err := CreateCurlPod(test, curlPodName, curlContainerName, namespace.Name)
+	g.Expect(err).NotTo(HaveOccurred())
+	// Wait until curl pod is created
+	g.Eventually(func(g Gomega) *corev1.Pod {
+		updatedCurlPod, err := test.Client().Core().CoreV1().Pods(curlPod.Namespace).Get(test.Ctx(), curlPod.Name, metav1.GetOptions{})
+		g.Expect(err).NotTo(HaveOccurred())
+		return updatedCurlPod
+	}, TestTimeoutShort).Should(WithTransform(sampleyaml.IsPodRunningAndReady, BeTrue()))
+
+	// test the default curl result
+	g.Eventually(func(g Gomega) {
+		// curl /fruit
+		stdout, _ := curlRayServicePod(test, rayService, curlPod, curlContainerName, "/fruit", `["MANGO", 2]`)
+		g.Expect(stdout.String()).To(Equal("6"))
+		// curl /calc
+		stdout, _ = curlRayServicePod(test, rayService, curlPod, curlContainerName, "/calc", `["MUL", 3]`)
+		g.Expect(stdout.String()).To(Equal("15 pizzas please!"))
+	}, TestTimeoutShort).Should(Succeed())
+
+	// After the above curl test, now the RayService is ready.
+	// We manually delete the "ray.io/serve" label on the Head for testing the reconciliation later.
+	g.Eventually(func(g Gomega) {
+		head, err := test.Client().Core().CoreV1().Pods(namespace.Name).Get(test.Ctx(), rayService.Status.ActiveServiceStatus.RayClusterStatus.Head.PodName, metav1.GetOptions{})
+		g.Expect(err).NotTo(HaveOccurred())
+		delete(head.Labels, utils.RayClusterServingServiceLabelKey)
+		_, err = test.Client().Core().CoreV1().Pods(namespace.Name).Update(test.Ctx(), head, metav1.UpdateOptions{})
+		g.Expect(err).NotTo(HaveOccurred())
+	}, TestTimeoutShort).Should(Succeed())
+
+	// upgrade the cluster and the serve application
+	// Parse ServeConfigV2 and replace the string in the simplest way to update it.
+	rayService, err = GetRayService(test, namespace.Name, rayService.Name)
+	g.Expect(err).NotTo(HaveOccurred())
+
+	serveConfig := rayService.Spec.ServeConfigV2
+	serveConfig = strings.Replace(serveConfig, "price: 3", "price: 4", -1)
+	serveConfig = strings.Replace(serveConfig, "factor: 5", "factor: 3", -1)
+
+	// modify EnableInTreeAutoscaling to trigger a zero downtime upgrade.
+	rayService.Spec.RayClusterSpec.EnableInTreeAutoscaling = ptr.To[bool](true)
```

Comment:
RayVersion seems to be a safer option.

Reply:
The newly added init container can already trigger the zero downtime upgrade.